testing/src/main/java/org/oppia/android/testing/espresso/TextInputAction.kt

```diff
@@ -0,0 +1,66 @@
+package org.oppia.android.testing.espresso
+
+import android.view.View
+import androidx.test.espresso.matcher.BoundedMatcher
+import com.google.android.material.textfield.TextInputLayout
+import org.hamcrest.Description
+
+/**
+ * Action for matching error text and performing other actions on [TextInputLayout] in a test
+ * infrastructure-specific way.
+ */
+class TextInputAction {
+
+  companion object {
+    /**
+     * Checks that the TextInputLayout view has the [expectedErrorText] error text.
+     *
+     * @returns a [BoundedMatcher] that matches the [expectedErrorText] with the TextInputLayout's
+     *     error text
+     */
+    fun hasErrorText(expectedErrorText: String): BoundedMatcher<View, TextInputLayout> {
+      return ErrorTextExisted(expectedErrorText)
+    }
+
+    /**
+     * Checks that the TextInputLayout view has an empty error text.
+     *
+     * @returns a [BoundedMatcher] that matches if error text is null or empty
+     */
+    fun hasNoErrorText(): BoundedMatcher<View, TextInputLayout> {
+      return ErrorTextNotExisted()
+    }
+
+    /**
+     * Class which inherits [BoundedMatcher] and overrides [matchesSafely] function to match the
+     * [expectedErrorText] with the TextInputLayout's error text.
+     *
+     */
+    class ErrorTextExisted(private val expectedErrorText: String) :
+      BoundedMatcher<View, TextInputLayout>(TextInputLayout::class.java) {
+      public override fun matchesSafely(textInputLayout: TextInputLayout): Boolean {
```

Comment:
Why is the public override needed? I think for your tests you shouldn't be interacting with these classes directly (in fact, they should be private). Instead, the test should be calling the public API of this file (which are the methods above) and interacting with those classes (which are of type BoundedMatcher). I think this actually simplifies your tests slightly & results in more robustness since they aren't being coupled with an implementation detail.

Reply:
PTAL now. Is this how you are saying to have these tests? But the only way I can have ```matchesSafely``` method being used in the test file is by interacting with the classes directly with the public override. Otherwise I am getting the following error while trying to call ```matchesSafely``` method : 
```Cannot access 'matchesSafely': it is protected in 'BoundedMatcher'```

Reply:
Can you try calling ``matches`` in your test, instead? That seems to be the public API for BoundedMatcher, and it calls matchesSafely() internally.

Reply:
This has been addressed in recent changes.